### DISCUSSION TITLE
SPIFFS to i2s Example

### DISCUSSION THREAD

**Discussion by craiglindley** [pschatzmann/arduino-audio-tools#31] [Q&A] [answered]
I noticed there wasn't an example using SPIFFS to i2s on the esp32. I thought I could modify the SD MP3 to i2s example easily enough but I couldn't get it to work. With my changes the code compiles and run but it stops immediately.

Any suggestions?

**Comment by craiglindley** [comment]
I think I know why. In the copy function of AudioCopy.h there is this line

            size_t len = available();

which returns 0 and therefore causes the code to return immediately. Shouldn't there be a stream on which available is called? Otherwise what is available referencing?

**Comment by pschatzmann** [comment]
I think the SPIFFS is a pain to use and not portable to other processors, therefore I did not provide any example.
The copy source needs to be a stream which has available data and if it is failing in your case you have an issue with the file content!

You can easily double check this by adding some asserts into the setup of your sketch:

```
  // setup file
  SPIFFS.begin();
  File audioFile = SPIFFS.open("/audio.mp3","r");
  assert(audioFile.available()>0);

```

**Reply by craiglindley** [reply]
Here is the code I use to open my file:

  // setup file
  if (! SPIFFS.begin(false)) {
    Serial.printf("Error starting SPIFFS\n");
    while (true) {}
  }
  File audioFile = SPIFFS.open("/HotelCAC.mp3", "r");
  if (! audioFile) {
    Serial.printf("Error opening file\n");
    while (true) {}
  }
  Serial.printf("File size: %d\n", audioFile.available());

Here is the results printed on the Serial interface

File size: 1621033

but internal to the copy function of AudioCopy.h

the available() method returns 0

There is a disconnect somewhere

**Comment by pschatzmann** [comment]
can you show the full sketch ?  
Maybe the issue is that audioFile is a local variable. Try to make it static or move the definition out of the method...

**Comment by craiglindley** [comment]
Yes you got it. When I make the audioFile variable global in the sketch the
file sizes match but now I get a corrupt heap error:

File size: 805091
len: 805091
CORRUPT HEAP: multi_heap.c:432 detected at 0x3ffb6090
abort() was called at PC 0x400891f3 on core 1

Since I used your sketch for SD to i2s as an example, I have to believe it
would have the same issues as my sketch has. That is, the available() in
the copy routine would return 0.

I have no idea what is causing the corrupt heap issue.

You can see the file size above and I have chosen 2mb app/2mb SPIFFS for
the partition scheme.

Sorry to keep bothering you about this but I would really like to be able
to code the app I have in mind.

On Mon, Nov 1, 2021 at 3:30 PM Phil Schatzmann ***@***.***>
wrote:

> can you show the full sketch ?
> Maybe the issue is that audioFile is a local variable. Try to make it
> static or move the definition out of the method...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pschatzmann/arduino-audio-tools/discussions/31#discussioncomment-1572761>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AH6K5D6J6SNOLAVFXDNRDKDUJ4BGPANCNFSM5HDBW3WQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Craig Lindley / Heather Hubbard

New Recordings are here <http://craigandheather.net/spellbound.html>

Personal Website is here <http://craigandheather.net>

Please call the Rockrimmon house. Our cell phones don't work at home.
Rockrimmon House: (719) 426-9873
Craig's Cell: (719) 502-7925
Heather's Cell: (719) 571-0944

If you’re one in a million, there are now more than seven thousand people
exactly like you.

**Comment by pschatzmann** [comment]
For some strange reason my sketch is working w/o issues - but I will correct it tomorrow. 
For the currupt heap I suggest that you update the libhelix library. I was getting the same issue on some mp3 files and increased the default output size...

**Comment by craiglindley** [comment, marked as answer]
Updating the library did it, thanks for the tip.

My only problem left is the audio still has some clicking in it which I
will investigate next.

On Mon, Nov 1, 2021 at 4:21 PM Phil Schatzmann ***@***.***>
wrote:

> For some strange reason my sketch is working w/o issues - but I will
> correct it tomorrow.
> For the currupt heap I suggest that you update the libhelix library. I was
> getting the same issue on some mp3 files and increased the default output
> size...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pschatzmann/arduino-audio-tools/discussions/31#discussioncomment-1572946>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AH6K5D6VMJZ5S3KTYOHINITUJ4HE7ANCNFSM5HDBW3WQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Craig Lindley / Heather Hubbard

New Recordings are here <http://craigandheather.net/spellbound.html>

Personal Website is here <http://craigandheather.net>

Please call the Rockrimmon house. Our cell phones don't work at home.
Rockrimmon House: (719) 426-9873
Craig's Cell: (719) 502-7925
Heather's Cell: (719) 571-0944

If you’re one in a million, there are now more than seven thousand people
exactly like you.

**Reply by JonS9999** [reply]
Hi Craig.  Were you ever able to resolve your clicking issue?  I'd also be curious to see your whole sketch (SPIFFS -> I2S).

Cheers.

**Comment by craiglindley** [comment]
No I was not able to resolve the issue so I gave up and went another way.

On Fri, Apr 8, 2022 at 1:24 PM JonS9999 ***@***.***> wrote:

> Hi Craig. Were you ever able to resolve your clicking issue? I'd also be
> curious to see your whole sketch (SPIFFS -> I2S).
>
> Cheers.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pschatzmann/arduino-audio-tools/discussions/31#discussioncomment-2533162>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AH6K5DY4CFZ76QRFS23ER2DVECBV5ANCNFSM5HDBW3WQ>
> .
> You are receiving this because you authored the thread.Message ID:
> <pschatzmann/arduino-audio-tools/repo-discussions/31/comments/2533162@
> github.com>
>


-- 
Craig Lindley / Heather Hubbard

New Recordings are here <http://craigandheather.net/spellbound.html>

Personal Website is here <http://craigandheather.net>

Please call the Rockrimmon house. Our cell phones don't work at home.
Rockrimmon House: (719) 426-9873
Craig's Cell: (719) 502-7925
Heather's Cell: (719) 571-0944

If you’re one in a million, there are now more than seven thousand people
exactly like you.

**Comment by pschatzmann** [comment]
I would expect that if you remove the meta data, the issue should go away

**Comment by craiglindley** [comment]
It didn't for me. Things got better but I can never make the clicking go
away completely

On Fri, Apr 8, 2022 at 1:30 PM Phil Schatzmann ***@***.***>
wrote:

> I would expect that if you remove the meta data, the issue should go away
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pschatzmann/arduino-audio-tools/discussions/31#discussioncomment-2533188>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AH6K5D4MXT5ZTYJIUQ3QQ63VECCONANCNFSM5HDBW3WQ>
> .
> You are receiving this because you authored the thread.Message ID:
> <pschatzmann/arduino-audio-tools/repo-discussions/31/comments/2533188@
> github.com>
>


-- 
Craig Lindley / Heather Hubbard

New Recordings are here <http://craigandheather.net/spellbound.html>

Personal Website is here <http://craigandheather.net>

Please call the Rockrimmon house. Our cell phones don't work at home.
Rockrimmon House: (719) 426-9873
Craig's Cell: (719) 502-7925
Heather's Cell: (719) 571-0944

If you’re one in a million, there are now more than seven thousand people
exactly like you.
